File: client/src/api/channels.ts
import { apiRequest } from "../services/api";
import type { Attachment } from "./media";
import type { ReactionSummary } from "./reactions";

export interface Channel {
  id: string;
  name: string;
  category: string | null;
  position: number;
  type: string;
}

export interface ThreadSummary {
  thread_id: string;
  reply_count: number;
  last_reply_at: string | null;
  last_reply_author_id: string | null;
}

export interface Message {
  id: string;
  channel_id: string;
  author_id: string;
  content: string;
  created_at: string;
  edited_at: string | null;
  deleted: boolean;
  attachments: Attachment[];
  reactions: ReactionSummary[];
  thread: ThreadSummary | null;
}

export interface MessagePage {
  messages: Message[];
  has_more: boolean;
}

export interface CreateChannelRequest {
  name: string;
  category?: string | null;
  position?: number;
}

export interface UpdateChannelRequest {
  name?: string;
  topic?: string | null;
  category?: string | null;
  position?: number;
}

export async function createChannel(
  baseUrl: string,
  token: string,
  req: CreateChannelRequest
): Promise<Channel> {
  return apiRequest<Channel>(baseUrl, "/api/v1/channels", {
    method: "POST",
    token,
    body: JSON.stringify(req),
  });
}

export async function updateChannel(
  baseUrl: string,
  token: string,
  channelId: string,
  req: UpdateChannelRequest
): Promise<Channel> {
  return apiRequest<Channel>(baseUrl, `/api/v1/channels/${channelId}`, {
    method: "PATCH",
    token,
    body: JSON.stringify(req),
  });
}

export async function deleteChannel(
  baseUrl: string,
  token: string,
  channelId: string
): Promise<void> {
  return apiRequest<void>(baseUrl, `/api/v1/channels/${channelId}`, {
    method: "DELETE",
    token,
  });
}

File: sdk/rust/src/rest/channels.rs
use crate::error::Result;
use crate::rest::models::{
    Channel, CreateChannelRequest, ListChannelsResponse, UpdateChannelRequest,
};
use crate::rest::Client;

pub struct ChannelsApi<'c> {
    client: &'c Client,
}

impl<'c> ChannelsApi<'c> {
    pub(crate) fn new(client: &'c Client) -> Self {
        Self { client }
    }

    pub async fn list(&self) -> Result<Vec<Channel>> {
        let resp: ListChannelsResponse = self.client.get("/api/v1/channels").await?;
        Ok(resp.channels)
    }

    pub async fn create(&self, req: &CreateChannelRequest) -> Result<Channel> {
        self.client.post("/api/v1/channels", req).await
    }

    pub async fn get(&self, channel_id: &str) -> Result<Channel> {
        self.client
            .get(&format!("/api/v1/channels/{channel_id}"))
            .await
    }

    pub async fn update(&self, channel_id: &str, req: &UpdateChannelRequest) -> Result<Channel> {
        self.client
            .patch(&format!("/api/v1/channels/{channel_id}"), req)
            .await
    }

    pub async fn delete(&self, channel_id: &str) -> Result<()> {
        self.client
            .delete(&format!("/api/v1/channels/{channel_id}"))
            .await
    }
}
